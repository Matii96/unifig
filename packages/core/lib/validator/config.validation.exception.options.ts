import { ValidationError } from 'class-validator';
import { Type } from '../utils/type.interface';

export interface ConfigValidationExceptionOptions {
  template: Type<any>;
  errors: ValidationError[];
}
