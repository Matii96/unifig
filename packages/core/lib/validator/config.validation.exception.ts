export class ConfigValidationException<TConfig extends object> extends Error {
  constructor(config: TConfig, msg: string) {
    super(`Validation failed for ${config.constructor.name}:${msg}`);
    this.name = ConfigValidationException.name;
  }
}
