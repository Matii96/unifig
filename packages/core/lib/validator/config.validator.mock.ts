export class ConfigValidatorMock {
  validate = jest.fn();
}
