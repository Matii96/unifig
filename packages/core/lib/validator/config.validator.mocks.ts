import { ValidationError } from 'class-validator';
import { DbConfigMock, TemplateMock } from '../core.mocks';

export const mockFailedValidation: ValidationError[] = [
  {
    target: new TemplateMock(),
    property: 'port',
    children: [] as any[],
    constraints: { isInt: 'port must be an integer number' },
  },
  {
    target: new TemplateMock(),
    value: {},
    property: 'db',
    children: [
      {
        target: new DbConfigMock(),
        property: 'url',
        children: [] as any[],
        constraints: { isString: 'url must be a string' },
      },
      {
        target: new DbConfigMock(),
        property: 'password',
        children: [],
        constraints: { isString: 'password must be a string' },
      },
    ],
  },
  {
    target: new TemplateMock(),
    value: {},
    property: 'db2',
    children: [
      {
        target: new DbConfigMock(),
        property: 'url',
        children: [] as any[],
        constraints: { isDefined: 'url must be a defined', isString: 'url must be a string' },
      },
      {
        target: new DbConfigMock(),
        property: 'subdb',
        children: [
          {
            target: new DbConfigMock(),
            property: 'url',
            children: [] as any[],
            constraints: { isString: 'url must be a string' },
          },
          {
            target: new DbConfigMock(),
            property: 'password',
            children: [],
            constraints: { isString: 'password must be a string' },
          },
        ],
        constraints: { isString: 'password must be a string' },
      },
    ],
  },
];
