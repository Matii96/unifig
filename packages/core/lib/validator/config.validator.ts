import { validateSync, ValidationError } from 'class-validator';
import { ConfigValidationException } from './config.validation.exception';

export class ConfigValidator {
  validate<TConfig extends object>(config: TConfig) {
    const errors = validateSync(config, { skipMissingProperties: false });
    if (errors.length === 0) {
      return;
    }
    const errorFormatted = '\n' + errors.map((error) => this.format(error)).join('');
    throw new ConfigValidationException(config, errorFormatted);
  }

  private format(error: ValidationError, parentPath = '') {
    let msg = error.constraints
      ? ` - ${parentPath}${error.property} has failed the following constraints:` +
        ` ${Object.keys(error.constraints).join(', ')}.` +
        ` Current value: ${error.value}\n`
      : '';
    msg += error.children?.map((child) => this.format(child, parentPath + `${error.property}.`)).join('');
    return msg;
  }
}
