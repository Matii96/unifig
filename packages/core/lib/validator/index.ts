export * from './decorators';
export * from './errors';
