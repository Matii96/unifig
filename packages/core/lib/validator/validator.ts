import { ConfigValidationError } from './errors/validation.error';

export interface Validator {
  validate(configs: object[]): ConfigValidationError | undefined;
}
