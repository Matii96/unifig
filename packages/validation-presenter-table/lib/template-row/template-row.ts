import {
  ConfigPropertyValidationError,
  ConfigSubtemplateValidationError,
  FailedConstraint,
} from '@unifig/core';

export class TemplateRow {
  propertyParentPrefix?: string;
  property: string;
  type?: string;
  source?: string;
  currentValue?: any;
  failedConstraints?: FailedConstraint[];

  static fromValidationError(
    error: ConfigPropertyValidationError | ConfigSubtemplateValidationError,
    propertyParentPrefix?: string
  ) {
    const row = new TemplateRow();
    row.propertyParentPrefix = propertyParentPrefix;
    row.property = error.property;
    row.type = error instanceof ConfigPropertyValidationError ? error.type : undefined;
    row.failedConstraints = error.failedConstraints;

    if (error instanceof ConfigPropertyValidationError) {
      row.source = error.source;
      row.currentValue = error.currentValue;
    }

    return row;
  }

  toArray(): [string, string, string, string, string, string] {
    return [
      '',
      (this.propertyParentPrefix ?? '') + this.property,
      this.type ?? '',
      this.source ?? '',
      this.formatValue(),
      this.failedConstraintsToNames(),
    ];
  }

  private failedConstraintsToNames() {
    return this.failedConstraints
      ? this.failedConstraints.map((failedConstraint) => failedConstraint.name).join(', ')
      : '';
  }

  private formatValue() {
    if (Number.isNaN(this.currentValue)) {
      return 'Not-A-Number';
    }
    return JSON.stringify(this.currentValue);
  }
}
