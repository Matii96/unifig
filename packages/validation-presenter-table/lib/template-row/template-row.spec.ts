import { mockFailed1dValidation, mockFailed2dValidation } from '../validator.mocks';
import { TemplateRow } from './template-row';

describe('TemplateRow', () => {
  it('should format row from 1d error', () => {
    expect(
      TemplateRow.fromValidationError(mockFailed1dValidation().errors[0]).toArray(),
    ).toMatchSnapshot();
  });

  it('should format table from 2d errors', () => {
    expect(
      TemplateRow.fromValidationError(mockFailed2dValidation()[0].errors[1]).toArray(),
    ).toMatchSnapshot();
  });
});
