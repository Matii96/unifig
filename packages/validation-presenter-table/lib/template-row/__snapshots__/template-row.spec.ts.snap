// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemplateRow should format row from 1d error 1`] = `
[
  "",
  "port",
  "Number",
  "PORT",
  ""not-a-port"",
  "isInt",
]
`;

exports[`TemplateRow should format table from 2d errors 1`] = `
[
  "",
  "db",
  "",
  "",
  undefined,
  "",
]
`;
